lib/core/mqtt.cpp

```diff
@@ -0,0 +1,164 @@
+#include "mqtt.hpp"
+
+#include <rapidjson/writer.h>
+
+#include "rapidjson/stringbuffer.h"
+
+namespace hyped::core {
+
+std::optional<std::shared_ptr<Mqtt>> Mqtt::create(ILogger &logger,
+                                                  const std::string &id,
+                                                  const std::string &host,
+                                                  const std::uint16_t port)
+{
+  std::string address = "tcp://" + host + ":" + std::to_string(port);
+  mqtt::connect_options connection_options;
+  connection_options.set_keep_alive_interval(1);
+  connection_options.set_clean_session(true);
+  auto mqtt_client = std::make_unique<mqtt::client>(address, id);
+  auto cb          = std::make_shared<MqttCallback>(logger);
+  mqtt_client->set_callback(*cb);
+  mqtt_client->connect(connection_options);
+  if (!mqtt_client->is_connected()) { return std::nullopt; }
+  return std::make_shared<Mqtt>(logger, std::move(mqtt_client), std::move(cb));
+}
+
+Mqtt::Mqtt(ILogger &logger, std::unique_ptr<mqtt::client> client, mqtt::callback_ptr cb)
+    : logger_(logger),
+      client_(std::move(client)),
+      messages_in_queue(0),
+      cb(std::move(cb))
+{
+}
+
+Mqtt::~Mqtt()
+{
+  client_->disconnect();
+}
+
+void Mqtt::publish(const MqttMessage &message, const MqttMessageQos qos)
+{
+  auto constructed_message = messageToMessagePtr(message);
+  logger_.log(core::LogLevel::kDebug,
+              "Publishing message %s",
+              constructed_message->get_payload_str().c_str());
+  constructed_message->set_qos(qos);
+  client_->publish(constructed_message);
+}
+
+core::Result Mqtt::subscribe(const core::MqttTopic topic)
+{
+  const auto topic_string = mqtt_topic_to_string.find(topic);
+  if (topic_string == mqtt_topic_to_string.end()) {
+    logger_.log(core::LogLevel::kFatal, "Attempted to subscribe to nonexistent MQTT topic");
+    return core::Result::kSuccess;
+  }
+  const auto result       = client_->subscribe(topic_string->second);
+  const auto reason_codes = result.get_reason_codes();
+  for (mqtt::ReasonCode code : reason_codes) {
+    if (code != mqtt::ReasonCode::SUCCESS && code != mqtt::ReasonCode::GRANTED_QOS_0
+        && code != mqtt::ReasonCode::GRANTED_QOS_1 && code != mqtt::ReasonCode::GRANTED_QOS_2) {
+      logger_.log(
+        core::LogLevel::kFatal, "Failed to subscribe to MQTT topic with error code %i", code);
+      return core::Result::kFailure;
+    }
+  }
+  logger_.log(core::LogLevel::kInfo, "Subscribed to topic");
+  return core::Result::kSuccess;
+}
+
+core::Result Mqtt::consume()
+{
+  mqtt::const_message_ptr received_msg;
+  auto received = client_->try_consume_message(&received_msg);
+  if (received) {
+    logger_.log(core::LogLevel::kInfo, "Received message");
+    // auto parsed_message = messagePtrToMessage(received_msg);
```

Comment:
Current problem is here, when we try and parse the received message there is a rapidjson related error